trieste/acquisition/rule.py

```diff
@@ -1103,6 +1115,380 @@ def state_func(
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> None:
+        """Update this box, including centre/location, using the given dataset. If the size of the
+        box is less than the minimum size, reinitialise the box."""
+        _, dataset = self.get_single_model_and_dataset(models, datasets)
+
+        if tf.reduce_any(self._eps < self._min_eps):
+            self.reinitialise(models, datasets)
+            return
+
+        _, y_min = self.get_local_min(dataset)
+
+        tr_volume = tf.reduce_prod(self.upper - self.lower)
+        step_is_success = y_min < self.y_min - self._kappa * tr_volume
+        self._eps = self._eps / self._beta if step_is_success else self._eps * self._beta
+
+        self._lower = tf.reduce_max(
+            [self.global_search_space.lower, self.location - self._eps], axis=0
+        )
+        self._upper = tf.reduce_min(
+            [self.global_search_space.upper, self.location + self._eps], axis=0
+        )
+
+        _, self.y_min = self.get_local_min(dataset)
+
+    def get_local_min(self, dataset: Optional[Dataset]) -> Tuple[TensorType, TensorType]:
+        """Calculate the local minimum of the box using the given dataset."""
+        if dataset is None:
+            raise ValueError("""dataset must be provided""")
+
+        in_tr = self.contains(dataset.query_points)
+        in_tr_obs = tf.where(
+            tf.expand_dims(in_tr, axis=-1),
+            dataset.observations,
+            tf.constant(np.inf, dtype=dataset.observations.dtype),
+        )
+        ix = tf.argmin(in_tr_obs)
+        x_min = tf.gather(dataset.query_points, ix)
+        y_min = tf.gather(in_tr_obs, ix)
+
+        return tf.squeeze(x_min, axis=0), tf.squeeze(y_min)
+
+
+def get_unique_points_mask(points: TensorType, tolerance: float = 1e-6) -> TensorType:
+    """Find unique points in a list, within a given tolerance."""
+
+    # Calculate the pairwise distances between points.
+    distances = tf.norm(tf.expand_dims(points, 1) - tf.expand_dims(points, 0), axis=-1)
+
+    # Create a mask that is True for the lower triangle of the distance matrix, not including
+    # the diagonal.
+    mask_tri = tf.linalg.band_part(tf.ones_like(distances, dtype=tf.bool), -1, 0)
+
+    # Use the mask to ignore the upper triangle and diagonal of the distance matrix.
+    distances_masked = tf.where(mask_tri, distances, tf.constant(np.inf, dtype=distances.dtype))
+
+    tolerance = tf.constant(tolerance, dtype=distances.dtype)
+
+    # Create a boolean mask for each point.
+    mask = tf.reduce_all(
+        tf.logical_or(
+            distances_masked > tolerance, tf.eye(distances_masked.shape[0], dtype=tf.bool)
+        ),
+        axis=1,
+    )
+
+    # Return the mask of unique points; can get the actual points with:
+    #   tf.boolean_mask(points, mask)
+    return mask
+
+
+class MultiTrustRegionBox(MultiTrustRegion[ProbabilisticModelType, TrustRegionBox]):
```

Comment:
I find this to be a bit of an overhead - say I want to implement a different type of reinit, e.g. I subclass the `TrustRegionBox` and change things, but then I also likely need to subclass this one with a bit of repetition in the init,  overwriting `create_subspace`...

Reply:
For significant changes to an algorithm (e.g. new parameters), yes it would require subclassing both.

To support simpler changes, e.g. to only the internals of `TrustRegionBox.reinitialize`, I can make a small tweak to `MultiTrustRegionBox` to use a new generic type instead of `TrustRegionBox`. This is exactly what the base `MultiTrustRegion` class does with `UpdateableSearchSpaceType`. It would avoid the need to override `create_subspace` and save subclassing in some cases.

Reply:
`create_subspace` has now been removed. The users are required to create the initial sequence of subspaces externally and pass them to the rule. This has the benefit that the users can decide which subspaces to create.